Assets/Scripts/Framework/Train/TrainController.cs

```diff
@@ -0,0 +1,271 @@
+using UnityEngine.Splines;
+using FishNet.Object;
+using UnityEngine;
+using FishNet;
+using System;
+
+namespace DerailedDeliveries.Framework.Train
+{
+    /// <summary>
+    /// Class responsible for moving train along rails spline.
+    /// </summary>
+    [RequireComponent(typeof(TrainEngine))]
+    public class TrainController : NetworkBehaviour
+    {
+        /// <summary>
+        /// Reference to the spline line data.
+        /// </summary>
+        [field: SerializeField]
+        public SplineContainer Spline { get; private set; }
+        
+        [Header("Train Config")]
+        [SerializeField, Range(0, 1)]
+        private float _trainFrontStartTime = 0f;
+
+        [SerializeField]
+        private float _heightOffset = 1.0f;
+
+        [Header("Wagons Config")]
+        [SerializeField]
+        private Transform _frontWagon = null;
+        
+        [SerializeField]
+        private float _wagonFollowDistance = 0f;
+
+        [SerializeField]
+        private float _wagonSpacing = 0f;
+
+        [SerializeField]
+        private Transform[] _wagons = null;
+
+        /// <summary>
+        /// Current distance value along spline lenght clamped between 0-1 (same as time). <br/>
+        /// <br/> 0 = Spline start point.<br/>
+        /// 1 = Spline end point.<br/>
+        /// </summary>
+        public float DistanceAlongSpline
+        {
+            get => _distanceAlongSpline;
+            set => _distanceAlongSpline = Mathf.Clamp01(value);
+        }
+
+        /// <summary>
+        /// Optimal start point for train on spline track based on its length. 
+        /// </summary>
+        public float CurrentOptimalStartPoint { get; private set; }
+
+        /// <summary>
+        /// Returns the precalculated line lenght of the spline
+        /// </summary>
+        public float SplineLength { get; private set; } = 0;
```

Comment:
This default value is also not needed

```suggestion
        public float SplineLength { get; private set; }
```